keps/sig-node/1977-container-notifier/README.md

```diff
@@ -0,0 +1,770 @@
+### Example Workflow with Sighup (Phase 2)
+
+This example involves a signal so it will be in phase 2 and beyond.
+
+For example, the user wants to send sighup to a container in a Pod. This signal is defined in the ContainterNotifierAction inside the Container.
+
+External controller creates a Notification object to request the sighup ContainerNotifier.
+Kubelet watches the Notification object and gets notified. It sends the sighup signal defined in the ContainerNotifierAction to the container. This is similar to "docker kill --signal=SIGHUP my_container".
+
+#### With Probe (Phase 3)
+
+If Probe is implement, Kubelet also sends a probe to check if the container is still running. If the probe detects that the container is stopped, Kubelet sets the Succeed field in ContainerNotificationStatus to True.
+If the container is still running, Kubelet will retry probes periodically until it times out. When that happens, it stops retrying and sets the ContainerNotificationStatus Succeed field to False.
+
+### Example Workflow to Change Log Verbosity (Phase 2)
+
+This example involves a signal so it will be in phase 2 and beyond.
+
+For example, the user wants to change log level to verbose in a container in a Pod. This signal is defined in the ContainterNotifierAction inside the Container.
+
+External controller creates a Notification object to request the ChangeLogLevel ContainerNotifier to change log to verbose.
+Kubelet watches the Notification object and gets notified. It sends the ChangeLogLevel to verbose signal defined in the ContainerNotifierAction to the container.
+
+#### With Probe (Phase 3)
+
+If Probe is implemented, Kubelet also sends a probe to check the log level in the container.
+
+If the probe detects that the log level inside the container is indeed verbose, Kubelet sets the Succeed field in ContainerNotificationStatus to True.
+If the probe detects that the log level is not changed yet, Kubelet will retry probes periodically until it times out. When that happens, it stops retrying and sets the ContainerNotificationStatus Succeed field to False.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Test Plan
+
+### Unit tests
+
+* Unit tests for container notifier controller will be added in phase 1.
+
+### E2E tests
+
+* E2e tests for creating a Notification object to request a ContainerNotifier when the feature flag is enabled.
+
+## Graduation Criteria
+
+#### Alpha Graduation
+
+* Feature Flag is present.
+* Container notifier controller is implemented in a sig-node sponsored repo.
+* E2E tests are implemented.
+
+### Alpha -> Beta Graduation
+
+* One release has been waited to allow for feedback from users.
```

Comment:
Are we willing to go Beta with the phase 1 design?  I don't think so, am I wrong?  I think you should talk about phases here.  If we decide we need phase 3, will we go beta without it?  What if we go beta and THEN decide we need phase 3, does that become alpha and we release independently?  Or do we wait for it to "catch up" and GA together.

Reply:
We'll need phase 2 before going Beta.  Regarding phase 3, we should discuss and make a decision after phase 2 is done?